### PR TITLE
filecache: log potential unlink error

### PR DESCRIPTION
**Pull request by sluongng**
We have received a report from a customer with self-hosted executor
regarding cleanup left over.

Let's start tracking if there is any failures during unlink syscall.

**Related issues**: N/A

### CODE REVIEW

File: enterprise/server/remote_execution/filecache/filecache.go [hunk 1]
```diff
@@ -90,12 +90,18 @@ func sizeFn(value interface{}) int64 {
...
-	if v, ok := value.(*entry); ok {
-		syscall.Unlink(v.value)
-		if reason == lru.SizeEviction {
-			age := time.Since(time.UnixMicro(v.addedAtUsec)).Microseconds()
-			metrics.FileCacheLastEvictionAgeUsec.Set(float64(age))
-		}
+	// TODO(sluongng): consider using Go generic to avoid type assertion.
+	v, ok := value.(*entry)
+	if !ok {
+		log.Warningf("could not cast filecache's value to entry struct: %v", value)
```
Comment (bduffany) [inline enterprise/server/remote_execution/filecache/filecache.go:96]:
```suggestion
		log.Errorf("Unexpected filecache entry type %T", value)
```

File: enterprise/server/remote_execution/filecache/filecache.go [hunk 2]
```diff
@@ -90,12 +90,18 @@ func sizeFn(value interface{}) int64 {
...
+		return
+	}
+
+	err := syscall.Unlink(v.value)
+	log.Errorf("failed to unlink filecache entry %q: %s", v.value, err)
```
Comment (bduffany) [inline enterprise/server/remote_execution/filecache/filecache.go:101]: (missing `err != nil` check)

```suggestion
  if err := syscall.Unlink(v.value); err != nil {
	  log.Errorf("Failed to unlink filecache entry %q: %s", v.value, err)
  }
```

File: enterprise/server/remote_execution/filecache/filecache.go [hunk 3]
```diff
@@ -90,12 +90,18 @@ func sizeFn(value interface{}) int64 {
 }
 
 func evictFn(value interface{}, reason lru.EvictionReason) {
-	if v, ok := value.(*entry); ok {
-		syscall.Unlink(v.value)
-		if reason == lru.SizeEviction {
-			age := time.Since(time.UnixMicro(v.addedAtUsec)).Microseconds()
-			metrics.FileCacheLastEvictionAgeUsec.Set(float64(age))
-		}
+	// TODO(sluongng): consider using Go generic to avoid type assertion.
```
Comment (bduffany) [inline enterprise/server/remote_execution/filecache/filecache.go:93]: I'm not sure this is the right place for this TODO - LRU is used pretty extensively in the codebase. I would omit this for now.

Comment (sluongng) [reply enterprise/server/remote_execution/filecache/filecache.go]: Agree

File: enterprise/server/remote_execution/filecache/filecache.go [hunk 4]
```diff
@@ -90,12 +90,18 @@ func sizeFn(value interface{}) int64 {
...
-	if v, ok := value.(*entry); ok {
-		syscall.Unlink(v.value)
-		if reason == lru.SizeEviction {
-			age := time.Since(time.UnixMicro(v.addedAtUsec)).Microseconds()
-			metrics.FileCacheLastEvictionAgeUsec.Set(float64(age))
-		}
+	v, ok := value.(*entry)
+	if !ok {
+		log.Errorf("Unexpected filecache entry type %T", value)
+		return
```
Comment (tylerwilliams) [inline enterprise/server/remote_execution/filecache/filecache.go:96]: you can use alert.UnexpectedEvent for stuff like this

**Comment by siggisim** [comment]
We can add this if you'd like, but I'm pretty sure the real issue here is https://github.com/buildbuddy-io/buildbuddy-internal/issues/2142

**Review by bduffany — APPROVED** [review]

**Pull request merged**